com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/persistence/query/impl/OperationQueryImpl.java

```diff
@@ -0,0 +1,194 @@
+                                                                       .attribute(AttributeNames.FINAL_STATE)
+                                                                       .condition(getCriteriaBuilder()::equal)
+                                                                       .value(finalState)
+                                                                       .build());
+        return this;
+    }
+
+    @Override
+    public OperationQuery startedBefore(Date startedBefore) {
+        queryCriteria.addRestriction(ImmutableQueryAttributeRestriction.<Date> builder()
+                                                                       .attribute(AttributeNames.STARTED_AT)
+                                                                       .condition(getCriteriaBuilder()::lessThan)
+                                                                       .value(startedBefore)
+                                                                       .build());
+        return this;
+    }
+
+    @Override
+    public OperationQuery endedAfter(Date endedAfter) {
+        queryCriteria.addRestriction(ImmutableQueryAttributeRestriction.<Date> builder()
+                                                                       .attribute(AttributeNames.ENDED_AT)
+                                                                       .condition(getCriteriaBuilder()::greaterThan)
+                                                                       .value(endedAfter)
+                                                                       .build());
+        return this;
+    }
+
+    @Override
+    public OperationQuery inNonFinalState() {
+        queryCriteria.addRestriction(ImmutableQueryAttributeRestriction.builder()
+                                                                       .attribute(AttributeNames.FINAL_STATE)
+                                                                       .condition(getCriteriaBuilder()::equal)
+                                                                       .value(null)
+                                                                       .build());
+        return this;
+    }
+
+    @Override
+    public OperationQuery inFinalState() {
+        queryCriteria.addRestriction(ImmutableQueryAttributeRestriction.builder()
+                                                                       .attribute(AttributeNames.FINAL_STATE)
+                                                                       .condition(getCriteriaBuilder()::notEqual)
+                                                                       .value(null)
+                                                                       .build());
+        return this;
+    }
+
+    @Override
+    public OperationQuery withStateAnyOf(List<State> states) {
+        queryCriteria.addRestriction(ImmutableQueryAttributeRestriction.<List<State>> builder()
```

Comment:
This is type inference for generic instance. What I pass to 'condition' method matters here. The 'lessThan' method signature requires a Comparable object. If I wasn't to specify 'Date' using the Diamond operator, it'd assume an Object, which is not Comparable, and it would not accept 'lessThan' as a valid condition.